vmware/ngc/NGC-Plugin/src/main/java/org/opensds/vmware/ngc/dao/impl/DeviceRepositoryImpl.java

```diff
@@ -0,0 +1,175 @@
+public class DeviceRepositoryImpl implements DeviceRepository {
+
+    private static final Log logger = LogFactory.getLog(DeviceRepositoryImpl.class);
+    private static Map<String, DeviceInfo> CACHE_DEVICE_INFO = new ConcurrentHashMap<>();
+    private static Map<String, Storage> LOGINED_DEVICE = new ConcurrentHashMap<>();
+    private static final String DEVICE_DATA_FILE_PATH = FileManager.getDeviceConfigPath() + Constant.DEVICE_DATA_FILE;
+
+    @PostConstruct
+    private void init() {
+        ObjectInputStream objectInputStream = null;
+        try {
+            File file = new File(DEVICE_DATA_FILE_PATH);
+            if (!file.exists()) {
+                return;
+            }
+            objectInputStream = new ObjectInputStream(new FileInputStream(file));
+            Map<String, DeviceInfo> deviceInfoMap = (Map<String, DeviceInfo>) objectInputStream.readObject();
+            reLogin(deviceInfoMap);
+            CACHE_DEVICE_INFO.putAll(deviceInfoMap);
+        } catch (Exception e) {
+            logger.warn(e.getMessage(),e);
+        } finally {
+            quietClose(objectInputStream);
+        }
+    }
+
+    @Override
+    public Map<String, DeviceInfo> getAll() {
+        return CACHE_DEVICE_INFO;
+    }
+
+    @Override
+    public void remove(String uid) {
+        DeviceInfo deviceInfo = CACHE_DEVICE_INFO.remove(uid);
+        LOGINED_DEVICE.remove(deviceInfo.ip);
+        writeDataToFile();
+    }
+
+    @Override
+    public void update(String uid, DeviceInfo deviceInfo) throws Exception {
+        initDevice(deviceInfo);
+        CACHE_DEVICE_INFO.put(uid, deviceInfo);
+        writeDataToFile();
+    }
+
+    @Override
+    public void add(String uid, DeviceInfo deviceInfo) throws Exception {
+        initDevice(deviceInfo);
+        CACHE_DEVICE_INFO.put(uid, deviceInfo);
+        writeDataToFile();
+    }
+
+
+    @Override
+    public DeviceInfo get(String uid) {
+        return CACHE_DEVICE_INFO.get(uid);
+    }
+
+    @Override
+    public Storage getLoginedDeviceByIP(String deviceIP) {
+        return LOGINED_DEVICE.get(deviceIP);
+    }
+
+
+    private void writeDataToFile() {
+        ObjectOutputStream objectOutputStream = null;
+        try {
+            File file = new File(DEVICE_DATA_FILE_PATH);
+            objectOutputStream = new ObjectOutputStream(new FileOutputStream(file));
+            synchronized (CACHE_DEVICE_INFO) {
+                objectOutputStream.writeObject(CACHE_DEVICE_INFO);
+            }
+        } catch (Exception e) {
+            logger.error(e.getMessage(),e);
+        } finally {
+            quietClose(objectOutputStream);
+        }
+    }
+
+    private void initDevice(DeviceInfo deviceInfo) throws Exception {
+        Storage device = StorageFactory.newStorage(getDeviceClassName(deviceInfo.deviceType), "");
+        device.login(deviceInfo.ip, deviceInfo.port, deviceInfo.username, deviceInfo.password);
+        StorageMO deviceMO = device.getDeviceInfo();
```

Comment:
relying on exceptions to be thrown, rather than error handlings, this may give tough time in debugging

Reply:
@kumarashit Do you mean Error handling by JVM? We are handling Error conditions for our programs using Exceptions. Do you have any other suggestions?